### PR TITLE
refactor: Add yyjson as Primary UniValue Storage Backend

### PR DESCRIPTION
**Pull request by pdath**
# refactor: Add yyjson as Primary UniValue Storage Backend

## Summary

This pull request introduces **yyjson** as an optional, high-performance primary storage backend for the UniValue library. yyjson is a battle-hardened, widely deployed JSON parser that is recognised as one of the fastest available JSON libraries.

**Key Features:**
- **Public API UniValue Compatible**: This implementation maintains compatibility with the existing UniValue library API, requiring **zero changes** to the rest of Bitcoin Knots codebase.
- **Disabled by default**, users must explicitly enable it using `-DWITH_YYJSON=ON`.  Existing users unaffected.
- **Performance Optimized**: Specifically designed to optimise Bitcoin Knots' performance for mining operations

## Architecture Overview

The implementation uses **yyjson as the primary storage mechanism** for all UniValue containers, with the original `val`, `keys`, and `values` members serving as lazy caches. This approach ensures that all UniValues benefit from yyjson's optimised serialisation, not just those parsed from JSON input.

### Semantics Change
Neither the original UniValue nor this yyjson version are thread-safe.  The caller must take precautions (such as using a mutex) to protect access.

This yyjson version is even less thread-safe.  A "const" UniValue parameter could potentially be changed by a different concurrent thread, breaking the "const" guarantee.  This limitation is accepted to maintain public API compatibility (and to avoid changes to anything else in Bitcoin Knots).  Everything in Bitcoin Knots **already** either uses a mutex to protect access to UniValue or is a per-request local, which prevents this from being an issue.

### Build System Integration

The implementation uses a CMake option `-DWITH_YYJSON` to toggle between the original UniValue implementation and the new yyjson-based implementation:

- **`-DWITH_YYJSON=OFF` (default)**: Uses the original UniValue files
- **`-DWITH_YYJSON=ON`**: Uses the new yyjson-based implementation

This dual-path approach enables zero-risk development and deployment.

### Code Organisation

To minimise duplication and maintain consistency, common methods shared between both implementations have been extracted to:
- `univalue_common.cpp`
- `univalue_common.h`

These common files contain approximately 99% identical code to the originals.  Several bugs resulting from failure to perform proper bounds checking were identified and fixed

The yyjson-specific implementation consists of:
- `univalue_yyjson.cpp` - Main yyjson backend implementation
- `univalue_yyjson_get.cpp` - Getter methods for yyjson backend
- `univalue_yyjson_read.cpp` - JSON reading/parsing with yyjson
- `univalue_yyjson_write.cpp` - JSON writing/serialization with yyjson
- `yyjson/yyjson.c` and `yyjson/yyjson.h` - The yyjson library itself

## Problem Being Solved

This pull request addresses the performance bottlenecks in Bitcoin Knots JSON serialisation and deserialization, particularly for mining-related operations. The standard UniValue implementation, while correct and reliable, has become a performance-limiting factor in high-throughput scenarios such as:

- Block to JSON conversion for mining
- Mempool RPC operations (`getblocktemplate`, `getblock`, etc.)
- Block template generation
- General JSON serialization/deserialization across the node

## Testing

Extensive testing has been performed on Ubuntu to ensure correctness, stability, and memory safety:
All tests have been run **with both `WITH_YYJSON=OFF` and `WITH_YYJSON=ON`**:

- All 57 existing UniValue unit tests (verifying API compatibility, behavioural compatibility, edge cases, error handling)
- All UniValue object tests  
- All ctest tests
- All ctest tests with Valgrind memory analysis
- All ctest tests with Address Sanitiser
- All functional tests

All tests pass cleanly with both the original and yyjson implementations.

I mined a block on [testnet4] against commit a103f93f40b50ee9842736d16c6bf6d37f1387d5: [143281](https://mempool.guide/testnet4/block/0000000000000000605413d9d439444748eee04a149f97c4dc62dafc7732da50)
I mined a block on [testnet4] against commit 2a3b19068678e8acf69e10d1e69018c10724fc1c: [144657](https://mempool.guide/testnet4/block/0000000000000001780bf4d6de53b8c2c7cab63bbf67ddbc721b4be7118e619f)

## Benchmark Results (on my machine)

Comprehensive benchmarking demonstrates significant improvements in RPC performance across various operations.  The test BlockToJsonVerboseWrite models the getblocktemplate function, and rpcmempool models mempool processing.
It is likely that there will be other flow-on benefits from reduced global mutex processing, but these have not been measured.

| Benchmark | OFF (ns) | ON (ns) | Improvement | Speedup |
|-----------|----------|---------|-------------|---------|
| BlockToJsonVerboseWrite | 36,976,768 | 7,258,135 | **+80.4%** | **5.1x** |
| BlockToJsonVerbosity1 | 542,752 | 530,374 | +2.3% | 1.0x |
| BlockToJsonVerbosity2 | 78,950,889 | 72,443,709 | +8.2% | 1.1x |
| BlockToJsonVerbosity3 | 78,781,516 | 72,291,663  | +8.2% | 1.1x |
| RpcMempool | 13,163,229 | 12,055,689 | +8.4% | 1.1x |

## Reproducing Benchmark Results

To reproduce the benchmark results, use the following commands.  Note that each benchmark will take 2.5 minutes to run.

## Build Without yyjson

```bash
cmake -B build-yyjson-off \
  -DRDTS_CONSENT=IMPLICIT \
  -DCMAKE_BUILD_TYPE=Release \
  -DBUILD_BENCH=ON

cmake --build build-yyjson-off -j$(nproc) -t bench_bitcoin 

./build-yyjson-off/bin/bench_bitcoin -filter="BlockToJson.*|RpcMempool" -min-time=30000 |  awk -F'|' '/,/{print $2,$NF}'
```

## Build With yyjson

```bash
cmake -B build-yyjson-on \
  -DRDTS_CONSENT=IMPLICIT \
  -DWITH_YYJSON=ON \
  -DCMAKE_BUILD_TYPE=Release \
  -DBUILD_BENCH=ON 

cmake --build build-yyjson-on -j$(nproc) -t bench_bitcoin 

./build-yyjson-on/bin/bench_bitcoin -filter="BlockToJson.*|RpcMempool" -min-time=30000 |  awk -F'|' '/,/{print $2,$NF}'
```

The above commands will output "ns/op" followed by the test name.
Compare the results to verify the performance improvements shown in the benchmark table above.

### CODE REVIEW

File: src/univalue/lib/univalue_yyjson.cpp [hunk 1]
```diff
@@ -0,0 +1,1292 @@
...
+        }
+
+        if (!use_legacy_path) {
+            yyjson_mut_val* new_val = nullptr;
+            if (val.m_yyjson_doc && val.m_yyjson_node) {
+                // val has its own yyjson tree - use yyjson's optimized copy function for mutable values
+                new_val = yyjson_mut_val_mut_copy(m_yyjson_doc.get(), val.m_yyjson_node);
+                if (!new_val) {
+                    // Copy failed, cannot add to array
+                    return;
```
Comment (kwsantiago) [inline src/univalue/lib/univalue_yyjson.cpp]: On allocation failure this returns without appending, and `m_materialized` is left unchanged, so the caller sees the old size and no error. The original throws `std::bad_alloc`. Under memory pressure an RPC response gets serialized with elements silently missing and reported as success.

Same pattern at :937, :1013, and both `pushKVEnd` paths. These should throw rather than drop.

File: src/univalue/lib/univalue_yyjson.cpp [hunk 2]
```diff
@@ -0,0 +1,1292 @@
...
+            m_materialized = true;
+        } else {
+            // Other has no yyjson state and is not materialized
+            // This shouldn't happen for valid objects, but handle it gracefully
+            keys = other.keys;
+            values = other.values;
+            m_materialized = true;
+        }
+        // For containers, we don't need yyjson state since we have materialized keys/values
+        m_yyjson_doc = nullptr;
```
Comment (kwsantiago) [inline src/univalue/lib/univalue_yyjson.cpp]: The copy constructor discards the yyjson tree for all containers. `push_back`/`pushKV` take `UniValue` by value, so passing an lvalue container copies it, strips its tree, hits `use_legacy_path` (:898-903, :990-995), and permanently resets the *parent's* doc (:955-957, :1058-1061). One lvalue push demotes the whole subtree to legacy storage for the object's lifetime.

Live example: `rpc/blockchain.cpp:1126`, `prune_lock_uv.pushKV("height", heights_uv)` with an lvalue VARR. `:1128` then demotes `locks_uv`.

The benchmark spread supports this. `core_write.cpp` moves throughout (:209, :218, :240, :261) and shows 5.35x on `BlockToJsonVerboseWrite`, while `BlockToJsonVerbosity2/3` sit at 1.08x. The fast path depends on every call site using `std::move`, and nothing enforces that.

File: src/univalue/lib/univalue_yyjson.cpp [hunk 3]
```diff
@@ -0,0 +1,1292 @@
...
+ *
+ * For primitives: Extracts the value from the yyjson node into `val`
+ * For arrays: Builds the `values` vector from the yyjson array
+ * For objects: Builds both `keys` and `values` vectors from the yyjson object
+ *
+ * Once materialized, subsequent accesses use the cached representation.
+ */
+void UniValue::materialize() const {
+    if (m_materialized) return;
+    if (!m_yyjson_doc || !m_yyjson_node) return;
```
Comment (kwsantiago) [inline src/univalue/lib/univalue_yyjson.cpp:682]: `materialize()` writes `typ`, `val`, `keys`, `values`, `m_yyjson_doc` and `m_materialized` through `const`, with no synchronization, from every read accessor. That removes the guarantee that a `const UniValue&` is safe to read concurrently.

Nothing in the tree relies on it today: `NullUniValue` is constructed materialized with a null doc (:101-106), so `materializeIfNeeded()` short-circuits, and every read of `ArgsManager::m_settings` holds `cs_args` (`common/args.cpp:450`, `:845`, `:881`, `:889`). So this is latent rather than live. But it is a silent contract change on a type passed as `const&` everywhere, and no CI job builds this flag, so neither the compiler nor TSan would catch the first violation.

Materialize eagerly, or guard with `call_once`, and document the contract either way.

File: src/univalue/lib/univalue.cpp [hunk 4]
```diff
@@ -134,13 +142,37 @@ void UniValue::pushKV(std::string key, UniValue val)
         pushKVEnd(std::move(key), std::move(val));
 }
 
-void UniValue::pushKVs(UniValue obj)
+void UniValue::pushKVs(const UniValue& obj)
```
Comment (kwsantiago) [inline src/univalue/lib/univalue.cpp]: This changes the default (non-yyjson) build. Callers passing `std::move(...)` (`wallet/rpc/addresses.cpp:478`, `rpc/output_script.cpp:85`) now bind to a const ref and copy instead of move.

Taking `obj` by value already made the aliasing guard unnecessary. This is unrelated to yyjson and should be dropped from the PR.

**Comment by kwsantiago** [comment]
The lint failure here is caused by this PR, not by pre-existing CI breakage.

Exactly one linter exits non-zero:

```
src/univalue/yyjson/yyjson.h seems to be missing the expected include guard:
  #ifndef BITCOIN_UNIVALUE_YYJSON_YYJSON_H
```

`yyjson.h` carries upstream's `#ifndef YYJSON_H`, and `src/univalue/yyjson/` is not in the lint exclusion list. The two failure markers in the log are one root cause: `lint-include-guards.py` fails, and `all_python_linters` re-runs the same script.

I checked out the PR head and ran every `test/lint/lint-*.py`. `lint-include-guards.py` is the only one that fails.

The rest of the lint log is non-fatal output that also appears on `29.x-knots`, which passes CI:

- `Duplicate include(s) in ...` — `lint-includes.py` prints these and exits 0.
- `FAIL: subtree directory was touched without subtree merge` for `src/minisketch` / `src/secp256k1` — printed, not fatal, and this PR touches neither.
- `circular-dependencies.py:29: unreachable code` — vulture output, prints `Python dead code detection found some issues` and continues.
- codespell `optin ==> option` — explicitly a warning.

On lint having been updated since the branch: the merge-base is `f41f01e1e6de`, which is the current `29.x-knots` tip, zero commits behind. The linters are in-tree, and that commit's own CI run passed.

Suggested fix, which keeps `yyjson.h` byte-identical to upstream 0.12.0:

```python
# test/lint/lint-include-guards.py
EXCLUDE_FILES_WITH_PREFIX = ['contrib/devtools/bitcoin-tidy',
                             'src/crypto/ctaes',
                             'src/tinyformat.h',
                             'src/bench/nanobench.h',
                             'src/univalue/yyjson/',
                             'src/test/fuzz/FuzzedDataProvider.h'] + SHARED_EXCLUDED_SUBTREES
```

With that one line, `lint-include-guards.py` exits 0 and all `test/lint/lint-*.py` pass on this branch.

Related: if `src/univalue/yyjson/` were registered as a subtree the way `secp256k1` and `minisketch` are, it would land in `SHARED_EXCLUDED_SUBTREES` automatically and there would be a documented path for pulling upstream fixes.

File: src/univalue/lib/univalue_yyjson.cpp (hunk 1 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp]: There are quite a few cases like this.  I will update them all to throw std::bad_alloc() and std::runtime_error() to more closely match the original UniValue.

File: src/univalue/lib/univalue.cpp (hunk 4 above)
Comment (pdath) [reply src/univalue/lib/univalue.cpp]: I'll reluctantly revert this one.  It passed all tests and provided a good performance gain.

**Comment by pdath** [comment]
> ```python
> # test/lint/lint-include-guards.py
> EXCLUDE_FILES_WITH_PREFIX = ['contrib/devtools/bitcoin-tidy',
>                              'src/crypto/ctaes',
>                              'src/tinyformat.h',
>                              'src/bench/nanobench.h',
>                              'src/univalue/yyjson/',
>                              'src/test/fuzz/FuzzedDataProvider.h'] + SHARED_EXCLUDED_SUBTREES
> ```
> 
> With that one line, `lint-include-guards.py` exits 0 and all `test/lint/lint-*.py` pass on this branch.

I have used your recommended fix.

File: src/univalue/lib/univalue_yyjson.cpp (hunk 3 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp:682]: I will use std::call_once.

File: src/univalue/lib/univalue_yyjson.cpp (hunk 2 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp]: This needed a lot deeper introspection.  I am going to fundamentally change the code to address this issue.

**Comment by pdath** [comment]
I have added commit 62cc41f5c8d41a73ded5af58e25b74f57c01b024 to address @kwsantiago comments.

Specifically addressed includes:
- Thread safety: Use std::atomic<bool> and std::mutex for lazy const access
  in materialization to support re-entry when materialized state changes
- Performance: Preserve yyjson tree through copy operations using
  yyjson_mut_val_mut_copy() to avoid performance regression from unconditional
  tree discarding
- Fix push_backV to handle nested aliases and prevent iterator invalidation
- Change pushKVs to take UniValue by value to avoid aliasing issues
- Remove unnecessary std::string allocation in UniValue::read
- Fix writeYyjson to not serialize entire document for child nodes
- Add vendor.md for yyjson version tracking
- Add Doxygen documentation for thread safety and performance considerations
- Add lint exclusion for yyjson vendor directory

File: src/univalue/include/univalue.h [hunk 5]
```diff
@@ -101,16 +151,64 @@ class UniValue {
...
+    mutable
+#endif
     std::vector<UniValue> values;
 
+#ifdef WITH_YYJSON
+    // yyjson primary storage
+    mutable std::shared_ptr<yyjson_mut_doc> m_yyjson_doc; //!< Shared pointer to yyjson mutable document (primary storage)
+    mutable yyjson_mut_val* m_yyjson_node{nullptr};    //!< Pointer to the root node in the yyjson tree
+    mutable std::atomic<bool> m_materialized{false};  //!< Whether lazy caches (val/keys/values) have been populated
+    mutable std::mutex m_materialize_mutex;          //!< Protects materialization to allow re-entry when m_materialized changes
```
Comment (kwsantiago) [inline src/univalue/include/univalue.h]: A `std::mutex` per UniValue is heavy: 40 bytes plus a non-trivial ctor/dtor on every instance, and UniValue is created per field and per array element. A `getblock`/`getblocktemplate` response builds and tears down thousands of them. A single doc-level lock, or the `std::call_once` you mentioned, would avoid paying this on every node.

File: src/univalue/lib/univalue_yyjson.cpp [hunk 6]
```diff
@@ -0,0 +1,1366 @@
...
+ * - Containers need documents to build their yyjson tree structure
+ * - Primitives don't need documents when they'll be pushed into containers
+ *
+ * @param type The value type (VOBJ, VARR, VNULL, VSTR, VNUM, or VBOOL)
+ * @param str The string value (for VSTR, VNUM, VBOOL) or ignored for containers
+ */
+UniValue::UniValue(UniValue::VType type, std::string str) : typ(type) {
+    // Containers need yyjson documents for tree building
+    if (type == VOBJ || type == VARR) {
+        m_yyjson_doc = std::shared_ptr<yyjson_mut_doc>(yyjson_mut_doc_new(nullptr), yyjson_doc_deleter);
```
Comment (kwsantiago) [inline src/univalue/lib/univalue_yyjson.cpp:127]: The copy ctor now allocates a fresh yyjson doc and deep-copies the tree on every container copy. That fixes the dropped-tree issue but copies are cheap in the legacy path, so this could be net slower. Worth re-running the benchmark table, it predates this commit and both this and the mutex change affect it.

File: src/univalue/lib/univalue_yyjson.cpp [hunk 7]
```diff
@@ -0,0 +1,1366 @@
...
+ * For arrays: Builds the `values` vector from the yyjson array
+ * For objects: Builds both `keys` and `values` vectors from the yyjson object
+ *
+ * Supports rematerialization when m_materialized is set to false (e.g., after
+ * push_back/pushKV add to the yyjson tree). Uses mutex for thread-safety.
+ *
+ * @note Thread-safe: Uses std::mutex to protect materialization and allow
+ * re-entry when the object transitions back to non-materialized state.
+ */
+void UniValue::materialize() const {
```
Comment (kwsantiago) [inline src/univalue/lib/univalue_yyjson.cpp]: Still no test and no CI job builds `-DWITH_YYJSON=ON`, so none of the thread-safety work is exercised. This change in particular needs a TSan test. Without a CI job the whole backend goes untested.

File: src/univalue/lib/univalue_yyjson.cpp (hunk 6 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp:127]: The current benchmark numbers in the top description have been re-run, but I will look into this further.

File: src/univalue/include/univalue.h (hunk 5 above)
Comment (pdath) [reply src/univalue/include/univalue.h]: I did attempt to use `std::call_once`, but I ran into an issue with rematerialization, and was forced to change.
The UniValue design requires re-running materialization when the yyjson tree changes.  Operations like push_back and pushKV modify the tree, then call "m_materialized.store(false);".

I like your idea of a single doc-level lock.  I plan to implement this idea.

File: src/univalue/lib/univalue_yyjson.cpp (hunk 7 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp]: I have never worked with GitActions.  My next commit adds this.  Could you please double-check it once it shows up.

File: src/univalue/include/univalue.h (hunk 5 above)
Comment (pdath) [reply src/univalue/include/univalue.h]: The single doc-level lock sounded so simple.  It has had a little complexity explosion.  :-)

File: src/univalue/lib/univalue_yyjson.cpp (hunk 3 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp:682]: I have done a U-turn on this, and materialised eagerly.  Using call_once/mutexs just didn't work out

File: src/univalue/include/univalue.h (hunk 5 above)
Comment (pdath) [reply src/univalue/include/univalue.h]: I have now removed all mutexes to match the original UniValue implementation.

File: src/univalue/lib/univalue_yyjson.cpp (hunk 7 above)
Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp]: I have now made the code match the original UniValue with respect to thread safety (both engines rely on the caller to assert a mutex).  Neither engine is thread-safe.

I could add a job to run all the Univalue unit and object style tests when ```-DWITH_YYJSON``` is enabled?  Your thoughts?

Comment (pdath) [reply src/univalue/lib/univalue_yyjson.cpp]: I have now added a CI test for x86 and arm64.

**Comment by mweinberg** [comment]
tACK 2a3b19068678e8acf69e10d1e69018c10724fc1c

Tested using four different AWS Graviton-based instances:

t4g.medium (2 x Graviton2 ARM)
m8g.medium (1 x Graviton4 ARM)
c8g.large (2 x Graviton4 ARM)
m9g.medium (1 x Graviton5 ARM)

## bench_bitcoin results — yyjson vs stock UniValue

_Bitcoin Knots v29.3.knots20260507 + pdath cherry-picks (yyjson, ARM crypto fix, jemalloc)_
_Values in nanoseconds per iteration (lower is better)._

### Compile flags

```
cmake -B build \
  -DENABLE_WALLET=OFF -DWITH_ZMQ=ON -DWITH_JEMALLOC=ON \
  -DWITH_YYJSON=ON \    # OFF for stock UniValue build
  -DENABLE_TOR_SUBPROCESS=OFF -U RDTS_CONSENT \
  -DBUILD_BENCH=ON -DBUILD_GUI=OFF -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_CXX_FLAGS_RELEASE="-O2 -g0 -march=armv8.2-a+crypto" \
  -DCMAKE_C_FLAGS_RELEASE="-O2 -g0 -march=armv8.2-a+crypto"
```

### Benchmark command

```
bench_bitcoin -filter="BlockToJson.*|RpcMempool" -min-time=30000
```

### yyjson ON

| Instance Type | CPU | VerboseWrite | Verbosity1 | Verbosity2 | Verbosity3 | RpcMempool |
|---|---|---:|---:|---:|---:|---:|
| m9g.medium | 1× Graviton5 | 2,648,488 | 346,482 | 26,621,077 | 26,367,255 | 4,640,323 |
| m8g.medium | 1× Graviton4 | 3,268,988 | 424,728 | 32,297,741 | 32,169,720 | 5,885,388 |
| c8g.large | 2× Graviton4 | 3,283,406 | 420,935 | 31,970,116 | 31,983,013 | 5,841,669 |
| t4g.medium | 2× Graviton2 | 6,305,617 | 691,679 | 59,804,034 | 59,750,441 | 11,033,945 |

### yyjson OFF (stock UniValue)

| Instance Type | CPU | VerboseWrite | Verbosity1 | Verbosity2 | Verbosity3 | RpcMempool |
|---|---|---:|---:|---:|---:|---:|
| m9g.medium | 1× Graviton5 | 19,948,626 | 350,061 | 28,734,921 | 28,586,425 | 4,007,772 |
| m8g.medium | 1× Graviton4 | 23,582,258 | 417,550 | 31,818,755 | 31,961,577 | 5,118,767 |
| c8g.large | 2× Graviton4 | 22,921,433 | 412,095 | 32,526,511 | 32,556,474 | 5,255,708 |
| t4g.medium | 2× Graviton2 | 38,520,707 | 651,211 | 61,295,996 | 62,122,416 | 11,089,145 |

### BlockToJsonVerboseWrite speedup (getblocktemplate path)

| Instance Type | CPU | Improvement | Speedup |
|---|---|---:|---:|
| m9g.medium | 1× Graviton5 | 86.7% | 7.53× |
| m8g.medium | 1× Graviton4 | 86.1% | 7.22× |
| c8g.large | 2× Graviton4 | 85.7% | 6.98× |
| t4g.medium | 2× Graviton2 | 83.6% | 6.11× |

### RpcMempool regression

| Instance Type | CPU | OFF (ns) | ON (ns) | Regression |
|---|---|---:|---:|---:|
| m9g.medium | 1× Graviton5 | 4,007,772 | 4,640,323 | -15.8% |
| m8g.medium | 1× Graviton4 | 5,118,767 | 5,885,388 | -15.0% |
| c8g.large | 2× Graviton4 | 5,255,708 | 5,841,669 | -11.1% |
| t4g.medium | 2× Graviton2 | 11,089,145 | 11,033,945 | +0.5% |

### Key findings

- **yyjson GBT speedup is consistent**: 6.1–7.5× across all Graviton generations
- **Graviton5 is fastest**: 2.6ms per GBT serialization (vs 3.3ms Graviton4, 6.3ms Graviton2)
- **Core count does not affect serialization benchmarks** (single-threaded)
- **RpcMempool regression is Graviton4/5 specific** (~11-16%); absent on Graviton2 — possibly related to memory allocation patterns on newer microarchitectures
- **Graviton2 → Graviton4 jump** is ~1.9× on VerboseWrite; **Graviton4 → Graviton5** is ~1.24×

File: src/univalue/include/univalue.h [hunk 8]
```diff
@@ -101,16 +148,65 @@ class UniValue {
...
+#ifdef WITH_YYJSON
+    mutable
+#endif
     std::vector<UniValue> values;
 
+#ifdef WITH_YYJSON
+    // yyjson primary storage
+    mutable std::shared_ptr<yyjson_mut_doc> m_yyjson_doc; //!< Shared pointer to yyjson mutable document (primary storage)
+    mutable yyjson_mut_val* m_yyjson_node{nullptr};    //!< Pointer to the root node in the yyjson tree
+    mutable bool m_materialized{false};  //!< Whether legacy caches (val/keys/values) have been populated
```
Comment (luke-jr) [inline src/univalue/include/univalue.h:189]: Why have any of the private members at all? Just store a yyjson pointer and remap the methods to call yyjson?

Comment (pdath) [reply src/univalue/include/univalue.h:189]: Because I am replacing an entire module, I decided early on that I wanted there to be zero changes to any tests or wider Bitcoin Knots code.  I also wanted there to be zero changes to the core battle-tested yyjson library (yyjson.c and yyjson.h).
I knew that if I were a reviewer, and I saw that someone had to change a bunch of tests to make their new code work, I would be very sceptical of the code's safety.

I discovered many little cases that failed the tests when I was initially passing everything through directly to yyjson.
For example, UniValue uses lowercase hex (\uXXXX), but yyjson uses uppercase hex.  The tests perform a case-sensitive match.
UniValue preserves leading zeros in front of a number (e,g. "020).  Yyjson does not.
So I had to write many little conversion functions to go from native yyjson to "UniValue" format.

I also wanted the yyjson variants of the files to loosely match the UniValue variants, to make it easier for those familiar with the existing UniValue code.  It should be intuitive where to look for the yyjson version.
But I needed some shared state like m_materialized between those files, so they ended up needing to be private.  Changes to any other files were unacceptable.

There are some wider changes I would be interested in making to the codebase that relate to improving the use of yyjson, but the performance gain for those doing mining is so large that I'd prefer to get this pull request accepted and on its long journey to being accepted into the codebase, and then work on the edge cases in the coming years, which are likely to provide minimal performance gains, but look cosmetically better.

I hope I understood the intent of your question correctly.

Comment (luke-jr) [reply src/univalue/include/univalue.h:189]: >I knew that if I were a reviewer, and I saw that someone had to change a bunch of tests to make their new code work, I would be very sceptical of the code's safety.

I appreciate that, but this _might_ go too far. (But I haven't seen the middle stages you're referring to, so I could be wrong)

>For example, UniValue uses lowercase hex (\uXXXX), but yyjson uses uppercase hex. The tests perform a case-sensitive match.

I think it's fine to adjust this assumption in tests.

> UniValue preserves leading zeros in front of a number (e,g. "020). Yyjson does not.

That sounds like a bug in UniValue! JSON doesn't allow a leading zero AIUI

Comment (pdath) [reply src/univalue/include/univalue.h:189]: I usually assume that if something produces a different output from yyjson, it is probably non-standard.  :-)

I'll do some research.  I need to look at software like DATUM gateway, ckpool, electrs, etc, and see if I can find anything with hard-coded assumptions about these edge cases.

This is an example of some of the massaging I had to do.  About 70 lines of "conversion" code instead of being able to just use the yyjson functions directly.

https://github.com/bitcoinknots/bitcoin/pull/326/changes#diff-1d5a6e687eb391ce10a0b118c788cb8970c592c7d84f3049dff1a69d708f4435R12

Comment (pdath) [reply src/univalue/include/univalue.h:189]: > Why have any of the private members at all? Just store a yyjson pointer and remap the methods to call yyjson?

@luke-jr , The JSON standard says that when encoding the DEL character, \u007f and 0x7f should be treated the same.  Interchangeable.

UniValue currently encodes DEL to \u007f.  yyjson encodes to 0x7f.
I currently have around 70 lines of code that takes already existing valid JSON containing a DEL character, and updates it to produce identical output to the original UniValue.  It is one of the private methods.

Would you rather the unitester treats \u007f and 0x7f the same (as per the json standard), or treat these as different so as to match existing UniValue behaviour?

I have had a look at DATUM gateway, ckpool, and electrs as a sample, and they all handle the JSON encoding either way.

**Review by Ari4ka — APPROVED** [review]

**Comment by delcin-raj** [comment]
cACK
